### ISSUE TITLE
Forgetting to call a variant constructor causes a confusing error message

### ISSUE THREAD

**Issue by arielb1** [rust-lang/rust#35241]
## STR

``` Rust
struct Foo(u32);

fn test() -> Foo { Foo }

fn main() {
}
```
## Confusing result

```
error[E0308]: mismatched types
 --> <anon>:3:20
  |
3 | fn test() -> Foo { Foo }
  |                    ^^^ expected struct `Foo`, found fn item
  |
  = note: expected type `Foo`
  = note:    found type `fn(u32) -> Foo {Foo::{{constructor}}}`
```

**Comment by arielb1** [comment]
The error message mentions function items, which is confusing to new users.

**Comment by estebank** [comment]
non-fn-like structs have better output, we should perform the same for fn-like structs:

```
error[E0423]: expected value, found struct `Bar`
 --> src/main.rs:9:21
  |
9 | fn test2() -> Bar { Bar }
  |                     ^^^ did you mean `Bar { /* fields */ }`?
```

**Comment by estebank** [comment]
### Mentoring instructions:

In [`note_type_err`](https://github.com/rust-lang/rust/blob/b7041bfab3a83702a8026fb7a18d8ea7d54cc648/src/librustc/infer/error_reporting/mod.rs#L669-L674), check wether `values` can be decomposed into `Some(ValuePairs::Types(exp_found)`. If it can, check wether `exp_found.found` is a [`TypeVariant::TyFnPtr(_)`](https://github.com/rust-lang/rust/blob/b7041bfab3a83702a8026fb7a18d8ea7d54cc648/src/librustc/ty/sty.rs#L129) with an `FnSig.output() == exp_found.expected`. If that is the case, add a `note` to the `diag` like for E-423.

**Comment by colinmarsh19** [comment]
Has anyone worked on this recently? I might take a crack at it. New to Rust but I'd like to learn, and start getting involved with OSS. Let me know if there are any conflicts, or if anyone is able to give me some guidance. Thanks! :)

**Comment by estebank** [comment]
@colinmarsh19 go ahead!

The terse explanation above should give enough context for someone slightly familiar with the compiler, but if you need more help you can always reach out with a comment in this ticket, on gitter or by creating a PR for us to comment on your code directly.

**Comment by joshleeb** [comment]
@colinmarsh19 are you currently working on this? If not I would like to give it a go :)

**Comment by colinmarsh19** [comment]
@joshleeb Yes, go ahead. Let me know of any progress you've made, I might look back later.

**Comment by dsheets** [comment]
I believe this issue has been fixed in #45630.
